sdk/identity/identity/src/msal/nodeFlows/msalClient.ts

```diff
@@ -366,20 +447,66 @@ To work with multiple accounts for the same Client ID and Tenant ID, please prov
     state.msalConfig.auth.clientCertificate = certificate;
 
     const msalApp = await getConfidentialApp(options);
-
-    return withSilentAuthentication(msalApp, scopes, options, () =>
-      msalApp.acquireTokenByClientCredential({
+    try {
+      const response = await msalApp.acquireTokenByClientCredential({
         scopes,
+        authority: state.msalConfig.auth.authority,
         azureRegion: calculateRegionalAuthority(),
+        claims: options?.claims,
+      });
+      ensureValidMsalToken(scopes, response, options);
+
+      msalLogger.getToken.info(formatSuccess(scopes));
+
+      return {
+        token: response.accessToken,
+        expiresOnTimestamp: response.expiresOn.getTime(),
+      };
+    } catch (err: any) {
+      throw handleMsalError(scopes, err, options);
+    }
+  }
+
+  async function getTokenByDeviceCode(
+    scopes: string[],
+    deviceCodeCallback: DeviceCodePromptCallback,
+    options: GetTokenWithSilentAuthOptions = {},
+  ): Promise<AccessToken> {
+    msalLogger.getToken.info(`Attempting to acquire token using device code`);
+
+    const msalApp = await getPublicApp(options);
+
+    return withSilentAuthentication(msalApp, scopes, options, () => {
+      const requestOptions: msal.DeviceCodeRequest = {
+        scopes,
+        cancel: false,
+        deviceCodeCallback,
         authority: state.msalConfig.auth.authority,
         claims: options?.claims,
-      }),
-    );
+      };
+      const deviceCodeRequest = msalApp.acquireTokenByDeviceCode(requestOptions);
+      if (options.abortSignal) {
+        options.abortSignal.addEventListener("abort", () => {
```

Comment:
I'm a bit confused here. Don't we need to check if the signal is already cancelled besides subscribing?

Also what good would setting the request options cancel property be after the options are passed into msal? Can we not cancel at that point?

Reply:
> I'm a bit confused here. Don't we need to check if the signal is already cancelled besides subscribing?

- Good catch, you're probably right! I can take another look

> Also what good would setting the request options cancel property be after the options are passed into msal? Can we not cancel at that point?

For this, and not that it's a good reason to do so, but my thought was to match the existing logic. msalDeviceCode follows the same pattern [here](https://github.com/Azure/azure-sdk-for-js/blob/455f2d5495bcfa71b5af2e1cfd0f80c848af1e16/sdk/identity/identity/src/msal/nodeFlows/msalDeviceCode.ts#L46-L51) and I wanted to match it. The MSAL docs aren't super clear on this but [the tests](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/d556aed39a401d45e45c17cd80f545f7cdc5f115/lib/msal-node/test/client/DeviceCodeClient.spec.ts#L565) do seem to follow a _similar_ pattern

Honestly, the best thing I could do (and should do) is test this behavior 😄 I'll try it and report back!

Reply:
To test this out, I updated the recorded test with the following code snippet:

```ts
const accessToken = await client.getTokenByDeviceCode(
        scopes,
        (info) => {
          console.log(
            `To complete the test recording, please go to ${info.verificationUri} and use code ${info.userCode} to authenticate.`,
          );
        },
        {
          abortSignal: AbortSignal.timeout(5000),
        },
      );
```

And can confirm that after 5 seconds the operation was aborted with the following message: `AbortError: The authentication has been aborted by the caller.`

I think because:

1. They pass the options bag through to the client in MSAL (so `cancel` param is a reference to the same param)
2. It's a long-poll operation

This works as expected and as the docs describe.

Regarding the first comment: Don't we need to check if the signal is already cancelled besides subscribing?

To test this, I passed the following:

```ts
      const accessToken = await client.getTokenByDeviceCode(
        scopes,
        (info) => {
          console.log(
            `To complete the test recording, please go to ${info.verificationUri} and use code ${info.userCode} to authenticate.`,
          );
        },
        {
          abortSignal: AbortSignal.abort(),
        },
      );
```

And you're right that we should check if the signal is already canceled - I'll push a fix for that shortly here! I think the only thing I need to do is set the initial value of `cancel` param to `true` if the signal is already canceled 👍